app/src/main/java/de/westnordost/streetcomplete/quests/building_entrance_reference/AddEntranceReference.kt

```diff
@@ -0,0 +1,95 @@
+import de.westnordost.streetcomplete.quests.building_entrance_reference.ReferenceCode
+import de.westnordost.streetcomplete.quests.building_entrance_reference.ReferenceCodeAndFlatRange
+import de.westnordost.streetcomplete.quests.building_entrance_reference.Unsigned
+
+class AddEntranceReference : OsmElementQuestType<EntranceAnswer> {
+
+    private val buildingFilter by lazy { """
+        ways with
+          building = apartments
+          and ruins != yes and disused != yes
+    """.toElementFilterExpression() }
+
+    override val changesetComment = "Add entrance identification"
+    override val wikiLink = "Key:ref"
+    override val icon = R.drawable.ic_quest_power // TODO!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+    override val questTypeAchievements = listOf(CITIZEN, BLIND)
+    override val enabledInCountries = NoCountriesExcept("BG", "PL", "RU", "UA", "BY", "EE")
+    override fun getTitle(tags: Map<String, String>) = R.string.quest_entrance_reference
+
+    override fun getHighlightedElements(element: Element, getMapData: () -> MapDataWithGeometry) =
+        getMapData().filter { it.tags.containsKey("entrance") }.asSequence()
+
+    override fun isApplicableTo(element: Element): Boolean? =
+        if (element !is Node || (element.tags["entrance"] != "main" && element.tags["entrance"] != "staircase") ) false else null
+
+    override fun getApplicableElements(mapData: MapDataWithGeometry): Iterable<Element> {
+        // note: it does not support multipolygon buildings
+        val buildingWays = mapData.ways.asSequence()
+            .filter { buildingFilter.matches(it) }
+        val returned = mutableListOf<Node>()
+
+        for (way in buildingWays) {
+            val entrancesForApartments = mutableListOf<Node>()
+            way.nodeIds.forEach { nodeId ->
+                val node = mapData.getNode(nodeId)
+                if(node != null) {
+                    if(node.tags["entrance"] == "main" || node.tags["entrance"] == "staircase") {
+                        entrancesForApartments.add(node)
+                    }
+                }
+            }
+            if(entrancesForApartments.count() >= 2) {
+                for (entrance in entrancesForApartments) {
+                    if(entrance.tags["noref"] == "yes") {
+                        continue
+                    }
+                    if(entrance.tags["ref:signed"] == "no") {
+                        continue
+                    }
+                    if(entrance.tags.containsKey("addr:flats")) {
+                        continue
+                    }
+                    if(entrance.tags.containsKey("name")) {
+                        continue
+                    }
+                    if(entrance.tags.containsKey("ref")) {
+                        continue
+                    }
+                    returned.add(entrance)
+                }
+            }
+        }
+        return returned
+    }
```

Comment:
Is it possible to refactor this without mutable lists to make it more concise and easier to read? Something like this:

```suggestion
    override fun getApplicableElements(mapData: MapDataWithGeometry): Iterable<Element> {
        // note: it does not support multipolygon buildings
        val buildingWays = mapData.ways.asSequence()
            .filter { buildingFilter.matches(it) }

        return buildingWays.flatMap { way ->
            val entrancesForApartments = way.nodeIds
                .mapNotNull { mapData.getNode(it) }
                .filter { it.tags["entrance"] == "main" || it.tags["entrance"] == "staircase" }

            if (entrancesForApartments.count() < 2) {
                return@flatMap emptyList()
            }

            return@flatMap entrancesForApartments.filter {
                it.tags["noref"] != "yes"
                && it.tags["ref:signed"] != "no"
                && !it.tags.containsKey("addr:flats")
                && !it.tags.containsKey("name")
                && !it.tags.containsKey("ref")
            }
        }
    }
```

Reply:
maybe but it still does not seem at all more readable to the current solution

or is there some other benefit?

Reply:
I think it's much more readable, but that is subjective of course. It should not have a (noticeable) impact on performance or change the logic, so there are no other benefits.

Reply:
For me at least map-reduce-filter code is not at all more readable.